ext/openssl/openssl_pwhash.c

```diff
@@ -0,0 +1,396 @@
+   | If you did not receive a copy of the PHP license and are unable to   |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@php.net so we can mail you a copy immediately.               |
+   +----------------------------------------------------------------------+
+   | Authors: Remi Collet <remi@php.net>                                  |
+   +----------------------------------------------------------------------+
+*/
+
+#ifdef HAVE_CONFIG_H
+# include "config.h"
+#endif
+
+#include "php.h"
+#include "ext/standard/php_password.h"
+#include "php_openssl.h"
+
+#if PHP_OPENSSL_API_VERSION >= 0x30200
+#include "Zend/zend_attributes.h"
+#include "openssl_pwhash_arginfo.h"
+#include <ext/standard/base64.h>
+#include "ext/random/php_random_csprng.h"
+#include <openssl/params.h>
+#include <openssl/core_names.h>
+#include <openssl/kdf.h>
+#include <openssl/thread.h>
+
+#define MEMLIMIT_MIN  8u
+#define MEMLIMIT_MAX  UINT32_MAX
+#define OPSLIMIT_MIN  1u
+#define OPSLIMIT_MAX  UINT32_MAX
+#define THREADS_MIN   1u
+#define THREADS_MAX   UINT32_MAX
+
+#define ARGON_VERSION 0x13
+
+#define SALT_SIZE     16
+#define HASH_SIZE     32
+#define DIGEST_SIZE   128
+
+static inline int get_options(zend_array *options, uint32_t *memlimit, uint32_t *opslimit, uint32_t *threads)
+{
+	zval *opt;
+
+	*opslimit = PHP_OPENSSL_PWHASH_OPSLIMIT;
+	*memlimit = PHP_OPENSSL_PWHASH_MEMLIMIT;
+	*threads  = PHP_OPENSSL_PWHASH_THREADS;
+
+	if (!options) {
+		return SUCCESS;
+	}
+	if ((opt = zend_hash_str_find(options, "memory_cost", strlen("memory_cost")))) {
+		zend_long smemlimit = zval_get_long(opt);
+
+		if ((smemlimit < 0) || (smemlimit < MEMLIMIT_MIN) || (smemlimit > (MEMLIMIT_MAX))) {
+			zend_value_error("Memory cost is outside of allowed memory range");
+			return FAILURE;
+		}
+		*memlimit = smemlimit;
+	}
+	if ((opt = zend_hash_str_find(options, "time_cost", strlen("time_cost")))) {
+		zend_long sopslimit = zval_get_long(opt);
+		if ((sopslimit < OPSLIMIT_MIN) || (sopslimit > OPSLIMIT_MAX)) {
+			zend_value_error("Time cost is outside of allowed time range");
+			return FAILURE;
+		}
+		*opslimit = sopslimit;
+	}
+	if ((opt = zend_hash_str_find(options, "threads", strlen("threads"))) && (zval_get_long(opt) != 1)) {
+		zend_long sthreads = zval_get_long(opt);
+		if ((sthreads < THREADS_MIN) || (sthreads > THREADS_MAX)) {
+			zend_value_error("Invalid number of threads");
+			return FAILURE;
+		}
+		*threads = sthreads;
+	}
+	return SUCCESS;
+}
+
+static bool php_openssl_argon2_compute_hash(
+	const char *algo,
+	uint32_t version, uint32_t memlimit, uint32_t opslimit, uint32_t threads,
+	const char *pass, size_t pass_len,
+	const unsigned char *salt, size_t salt_len,
+	unsigned char *hash, size_t hash_len)
+{
+	OSSL_PARAM params[7], *p = params;
+	EVP_KDF *kdf = NULL;
+	EVP_KDF_CTX *kctx = NULL;
+	bool ret = false;
+
+
+	if (threads > 1) {
+		if (OSSL_set_max_threads(NULL, threads) != 1) {
```

Comment:
This might be problematic for ZTS. It seems like a global option which might not play nicely when used by different threads - at least it needs some checking. We might need to look to introduction per worker thread lib ctx. Coincidentally, this is something that I'm starting to look due to another (FrankenPHP) related issue with openssl_encrypt.

Reply:
> This might be problematic for ZTS. It seems like a global option which might not play nicely when used by different threads - at least it needs some checking. We might need to look to introduction per worker thread lib ctx. Coincidentally, this is something that I'm starting to look due to another (FrankenPHP) related issue with openssl_encrypt.

Do you want me to create a local ctx in this function ? or do you prefer to keep as it is, and wait for your work on a per thread lib ctx ?

Reply:
I don't think it should be done in this way. Context should be ideally global (it means re-usable for each thread) but there are couple of other things to consider. I will be looking into it in the coming weeks and see. Will ping you when I got something sensible.